### PR TITLE
[REVIEW] Create cudf::detail::byte_cast for cudf::byte_cast

### PR DESCRIPTION
**Pull request by davidwendt**
Found this when I noticed a doxygen error in `reshape.hpp`
https://github.com/rapidsai/cudf/blob/47a3d32cc343293c5240c1c60f0dcd3ef7e26691/cpp/include/cudf/reshape.hpp#L75
This line should be right before the end bracket for the `namespace cudf` line so the `cudf::byte_cast` appears in the correct doxygen group.

Noticed that `cudf::byte_cast` includes a `cudaStream_t` parameter unlike all the other APIs in this header file. This PR creates the appropriate `cudf::detail::byte_cast` and move the `cudaStream_t` parameter to that call per the coding guidelines here:
https://github.com/rapidsai/cudf/blob/branch-0.17/cpp/docs/TRANSITIONGUIDE.md#streams

### CODE REVIEW

**Review by codereport — APPROVED** [review]
LGMT :+1:

File: cpp/src/reshape/byte_cast.cu [hunk 1]
```diff
@@ -107,9 +108,23 @@ std::unique_ptr<column> byte_cast(column_view const& input_column,
...
   return type_dispatcher(
     input_column.type(), byte_list_conversion{}, input_column, configuration, mr, stream);
 }
 
+}  // namespace detail
+
+/**
+ * @copydoc cudf::byte_cast(input_column,flip_endianess,rmm::mr::device_memory_resource)
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
```
Comment (harrism) [inline cpp/src/reshape/byte_cast.cu:120]: No stream here.

```suggestion
```

File: cpp/src/reshape/byte_cast.cu [hunk 2]
```diff
@@ -107,9 +108,23 @@ std::unique_ptr<column> byte_cast(column_view const& input_column,
...
+ */
+std::unique_ptr<column> byte_cast(column_view const& input_column,
+                                  flip_endianness configuration,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  return detail::byte_cast(input_column, configuration, mr, 0);
```
Comment (harrism) [inline cpp/src/reshape/byte_cast.cu:127]:
```suggestion
  return detail::byte_cast(input_column, configuration, mr, cudaStreamDefault);
```

**Review by harrism — CHANGES_REQUESTED** [review]
Couple of doc improvements and a parameter name improvement.

**Review by harrism — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Pull request merged**